### PR TITLE
Qcut sampling postprocessing function

### PR DESCRIPTION
**Pull request by anthayes92**
**Context:**
The pipeline for cutting a circuit with sampling measurements requires it own postprocessing capabilities. This will support returning bitstrings as well as expectation values, the choice of which should be specified by the user. 

**Description of the Change:**
Add sampling postprocessing function

**Benefits:**
Facilitates cutting circuits with sampling measurements

**Possible Drawbacks:**
Potential computational bottleneck depending on number of fragments circuits to be processed

### CODE REVIEW

**Comment by anthayes92** [comment]
[sc-15890]

File: tests/transforms/test_qcut.py [hunk 1]
```diff
@@ -1959,6 +1962,79 @@ def test_mc_state_prep(self):
                 assert op.wires == exp_op.wires
 
 
+class TestMCPostprocessing:
+    """
+    Tests that the postprocessing for circuits containing sample measurements
+    gives the correct results.
+    """
+
+    def test_sample_postprocess(self):
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1971]: Might be good to have an interface-based test like [here](https://github.com/PennyLaneAI/pennylane/blob/482cd0894548adeb1d507d7debf84aa321cc0726/tests/transforms/test_qcut.py#L2212), where we also check that the returned is converted to the right interface type.

File: tests/transforms/test_qcut.py [hunk 2]
```diff
@@ -1959,6 +1962,79 @@ def test_mc_state_prep(self):
...
+
+        postprocessed = qcut.qcut_processing_fn_mc(
+            fixed_samples, communication_graph, fixed_settings, shots, func
+        )
+
+        expected = 85.33333333333333
+
+        assert np.isclose(postprocessed, expected)
+
+
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:2037]: Perhaps worth adding a quick test for `_reshape_and_find_degrees`.

**Review by trbromley — CHANGES_REQUESTED** [review]
Thanks @anthayes92!

File: pennylane/transforms/qcut.py [hunk 3]
```diff
@@ -827,6 +827,116 @@ def expand_fragment_tapes_mc(
...
+
+        assert set(sample_terminal).issubset({np.array(0), np.array(1)})
+        assert set(sample_mid).issubset({np.array(-1), np.array(1)})
+        # following Eq.(35) of Peng et.al: https://arxiv.org/abs/1904.00102
+        f = classical_processing_fn(sample_terminal)
+        sigma_s = np.prod(sample_mid)
+        t_s = f * sigma_s
+        c_s = np.prod([evals[s] for s in setting])
+        K = len(sample_mid)
+        expvals.append(8**K * c_s * t_s)
```
Comment (angusjlowe) [inline pennylane/transforms/qcut.py:935]:
```suggestion
    vals = []
    for result, setting in zip(results, settings.T):
        sample_terminal = []
        sample_mid = []

        for fragment_result, out_degree in zip(result, out_degrees):
            sample_terminal.append(fragment_result[: -out_degree or None])
            sample_mid.append(fragment_result[-out_degree or len(fragment_result) :])

        sample_terminal = np.hstack(sample_terminal)
        sample_mid = np.hstack(sample_mid)

        assert set(sample_terminal).issubset({np.array(0), np.array(1)})
        assert set(sample_mid).issubset({np.array(-1), np.array(1)})
        # following Eq.(35) of Peng et.al: https://arxiv.org/abs/1904.00102
        f = classical_processing_fn(sample_terminal)
        sigma_s = np.prod(sample_mid)
        t_s = f * sigma_s
        c_s = np.prod([evals[s] for s in setting])
        K = len(sample_mid)
        vals.append(8**K * c_s * t_s)
```

File: tests/transforms/test_qcut.py [hunk 4]
```diff
@@ -1959,6 +1962,157 @@ def test_mc_state_prep(self):
...
+            if x[0] == 0 and x[1] == 1:
+                return -1
+            if x[0] == 1 and x[1] == 0:
+                return -1
+            if x[0] == 1 and x[1] == 1:
+                return 1
+
+        fixed_settings = np.array([[0, 7, 1], [5, 7, 2], [1, 0, 3], [5, 1, 1]])
+
+        # spy_func = mocker.spy(fn)
```
Comment (anthayes92) [inline tests/transforms/test_qcut.py:2038]: I'm not sure how to spy on this function, generally we use `mocker.spy(package, "function")` but this function exists by itself so I only have a single arg to pass to spy

File: tests/transforms/test_qcut.py (hunk 2 above)
Comment (anthayes92) [reply tests/transforms/test_qcut.py:2115]: Added thanks!

File: pennylane/transforms/qcut.py [hunk 5]
```diff
@@ -827,6 +827,118 @@ def expand_fragment_tapes_mc(
...
+    results = [qml.math.flatten(r) for r in results]
+    results = np.array(results, dtype=object)
+    results = results.reshape((len(communication_graph), shots)).T
+
+    return results
+
+
+def qcut_processing_fn_sample(
+    results: Sequence, communication_graph: MultiDiGraph, shots: int
+) -> List[np.array]:
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:844]:
```suggestion
) -> List:
```
This could be e.g. a TF tensor

File: tests/transforms/test_qcut.py [hunk 6]
```diff
@@ -1959,6 +1962,157 @@ def test_mc_state_prep(self):
...
+            np.array([[1.0], [0.0], [1.0], [1.0]]),
+            np.array([[0.0], [0.0], [1.0], [-1.0]]),
+            np.array([[0.0], [1.0], [1.0], [-1.0]]),
+            np.array([[0.0], [-1.0], [1.0]]),
+            np.array([[0.0], [-1.0], [-1.0]]),
+            np.array([[1.0], [1.0], [1.0]]),
+        ]
+
+        postprocessed = qcut.qcut_processing_fn_sample(fixed_samples, communication_graph, shots)
+        postprocessed = qml.math.cast_like(postprocessed, lib.ones(1))
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1996]: Why are we casting the output to the interface type? We need to cast the input to the interface type and check that the output is of the same type.

Comment (trbromley) [reply tests/transforms/test_qcut.py:1996]: (same for the test below)

File: tests/transforms/test_qcut.py (hunk 4 above)
Comment (trbromley) [reply tests/transforms/test_qcut.py:2038]: Ok, can leave out for now :+1:

File: tests/transforms/test_qcut.py [hunk 7]
```diff
@@ -1959,6 +1962,157 @@ def test_mc_state_prep(self):
...
+
+        edge_data = [(0, 1, {"pair": (qcut.MeasureNode(wires=[1]), qcut.PrepareNode(wires=[1]))})]
+        communication_graph = MultiDiGraph(edge_data)
+        shots = 5
+
+        reshaped = qcut._reshape_results(results, communication_graph, shots)
+
+        for resh, exp_resh in zip(reshaped, expected_reshaped):
+            for arr, exp_arr in zip(resh, exp_resh):
+                assert np.allclose(arr, exp_arr)
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:2113]: Does a big `np.allclose(reshaped, expected_reshaped)` not work because the arrays contain objects?

**Review by trbromley — COMMENTED** [review]
Thanks @anthayes92! Just left a few more comments - would it also be possible to fix what is mentioned [here](https://github.com/PennyLaneAI/pennylane/pull/2358#discussion_r833538455)?

File: pennylane/transforms/qcut.py [hunk 8]
```diff
@@ -827,6 +827,118 @@ def expand_fragment_tapes_mc(
...
+        sample_mid = []
+
+        for fragment_result, out_degree in zip(result, out_degrees):
+            sample_terminal.append(fragment_result[: -out_degree or None])
+            sample_mid.append(fragment_result[-out_degree or len(fragment_result) :])
+
+        sample_terminal = np.hstack(sample_terminal)
+        sample_mid = np.hstack(sample_mid)
+
+        assert set(sample_terminal).issubset({np.array(0), np.array(1)})
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py:926]: yes that's what I found!

File: tests/transforms/test_qcut.py (hunk 7 above)
Comment (anthayes92) [reply tests/transforms/test_qcut.py:2116]: exactly, also why there is a second for loop

**Comment by anthayes92** [comment]
Thanks for the suggestions @trbromley , I've updated the docsting for `expand_fragment_tapes_mc` to include the `settings` array and also changed the tape diagrams since the `tape.draw()` function has been updated to give cleaner output.

File: pennylane/transforms/qcut.py [hunk 9]
```diff
@@ -827,6 +824,118 @@ def expand_fragment_tapes_mc(
...
+    transform. This removes superfluous mid-circuit measurement samples from fragment
+    circuit outputs.
+
+    .. note::
+
+        This function is designed for use as part of the sampling-based circuit cutting workflow.
+        Check out the :func:`qml.cut_circuit_mc() <pennylane.cut_circuit_mc>` transform for more details.
+
+    Args:
+        results (Sequence): a collection of execution results generated from the
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:853]:
```suggestion
        results (Sequence): a collection of sample-based execution results generated from the
```

File: pennylane/transforms/qcut.py [hunk 10]
```diff
@@ -827,6 +824,118 @@ def expand_fragment_tapes_mc(
...
+    transform. This takes a user-specified classical function to act on bitstrings and
+    generates an expectation value.
+
+    .. note::
+
+        This function is designed for use as part of the sampling-based circuit cutting workflow.
+        Check out the :func:`qml.cut_circuit_mc() <pennylane.cut_circuit_mc>` transform for more details.
+
+    Args:
+        results (Sequence): a collection of execution results generated from the
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:893]:
```suggestion
        results (Sequence): a collection of sample-based execution results generated from the
```

File: pennylane/transforms/qcut.py [hunk 11]
```diff
@@ -827,6 +824,118 @@ def expand_fragment_tapes_mc(
...
+            random expansion of circuit fragments over measurement and preparation node configurations
+        communication_graph (nx.MultiDiGraph): the communication graph determining connectivity
+            between circuit fragments
+        settings (tensor_like): each element is one of 8 unique values that determines and tracks the specific
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:897]:
```suggestion
        settings (np.ndarray): Each element is one of 8 unique values that tracks the specific
```
We have control over the type of `settings`.

File: pennylane/transforms/qcut.py [hunk 12]
```diff
@@ -827,6 +824,118 @@ def expand_fragment_tapes_mc(
...
+            measurement and preparation operations over all configurations. The number of rows is determined
+            by the number of cuts and the number of columns is determined by the number of shots.
+        shots (int): the number of shots
+        classical_processing_fn (callable): a classical postprocessing function to be applied to
+            the reconstructed bitstrings. The expected input is a bitstring; a flat array of length ``wires``
+            and the output should be a single number within the interval [-1, 1]
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:903]:
```suggestion
            and the output should be a single number within the interval :math:`[-1, 1]`.
```

File: pennylane/transforms/qcut.py [hunk 13]
```diff
@@ -827,6 +824,118 @@ def expand_fragment_tapes_mc(
...
+    Args:
+        results (Sequence): a collection of execution results generated from the
+            random expansion of circuit fragments over measurement and preparation node configurations
+        communication_graph (nx.MultiDiGraph): the communication graph determining connectivity
+            between circuit fragments
+        settings (tensor_like): each element is one of 8 unique values that determines and tracks the specific
+            measurement and preparation operations over all configurations. The number of rows is determined
+            by the number of cuts and the number of columns is determined by the number of shots.
+        shots (int): the number of shots
+        classical_processing_fn (callable): a classical postprocessing function to be applied to
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:901]:
```suggestion
        classical_processing_fn (callable): A classical postprocessing function to be applied to
```

File: tests/transforms/test_qcut.py [hunk 14]
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
...
+        correct result
+        """
+        lib = pytest.importorskip(interface)
+        fragment_tapes = [frag0, frag1]
+
+        communication_graph = MultiDiGraph(frag_edge_data)
+        shots = 3
+        fragment_configurations, settings = qml.transforms.qcut.expand_fragment_tapes_mc(
+            fragment_tapes, communication_graph, shots
+        )
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1984]:
```suggestion
```
Do we need this?

File: tests/transforms/test_qcut.py [hunk 15]
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
...
+
+        fixed_samples = [
+            np.array([[1.0], [0.0], [1.0], [1.0]]),
+            np.array([[0.0], [0.0], [1.0], [-1.0]]),
+            np.array([[0.0], [1.0], [1.0], [-1.0]]),
+            np.array([[0.0], [-1.0], [1.0]]),
+            np.array([[0.0], [-1.0], [-1.0]]),
+            np.array([[1.0], [1.0], [1.0]]),
+        ]
+        cast_fixed_samples = [qml.math.cast_like(fs, lib.ones(1)) for fs in fixed_samples]
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1994]:
```suggestion
        cast_fixed_samples = [qml.math.convert_like(fs, lib.ones(1)) for fs in fixed_samples]
```
This is an important distinction.

File: tests/transforms/test_qcut.py (hunk 14 above)
Comment (trbromley) [reply tests/transforms/test_qcut.py:1984]: BTW, not sure what IDE you are using, but on PyCharm this is easy to see with unused variables shown in a faded grey colour:
image

Maybe something that can be enabled?

File: tests/transforms/test_qcut.py [hunk 16]
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
...
+        correct result
+        """
+        lib = pytest.importorskip(interface)
+        fragment_tapes = [frag0, frag1]
+
+        communication_graph = MultiDiGraph(frag_edge_data)
+        shots = 3
+        fragment_configurations, settings = qml.transforms.qcut.expand_fragment_tapes_mc(
+            fragment_tapes, communication_graph, shots
+        )
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:2017]:
```suggestion
```

File: tests/transforms/test_qcut.py [hunk 17]
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
...
+
+        fixed_samples = [
+            np.array([[1.0], [0.0], [1.0], [1.0]]),
+            np.array([[0.0], [0.0], [1.0], [-1.0]]),
+            np.array([[0.0], [1.0], [1.0], [-1.0]]),
+            np.array([[0.0], [-1.0], [1.0]]),
+            np.array([[0.0], [-1.0], [-1.0]]),
+            np.array([[1.0], [1.0], [1.0]]),
+        ]
+        cast_fixed_samples = [qml.math.cast_like(fs, lib.ones(1)) for fs in fixed_samples]
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:2027]:
```suggestion
        cast_fixed_samples = [qml.math.convert_like(fs, lib.ones(1)) for fs in fixed_samples]
```

File: tests/transforms/test_qcut.py [hunk 18]
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
...
+
+        fixed_settings = np.array([[0, 7, 1], [5, 7, 2], [1, 0, 3], [5, 1, 1]])
+
+        spy_prod = mocker.spy(np, "prod")
+        spy_hstack = mocker.spy(np, "hstack")
+
+        postprocessed = qcut.qcut_processing_fn_mc(
+            fixed_samples, communication_graph, fixed_settings, shots, fn
+        )
+        postprocessed = qml.math.cast_like(postprocessed, lib.ones(1))
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:2047]:
```suggestion
```
I don't think we need this - we should be testing that `postprocessed` is the expected type without casting.

File: tests/transforms/test_qcut.py [hunk 19]
```diff
@@ -1959,6 +1962,160 @@ def test_mc_state_prep(self):
...
+            if x[0] == 1 and x[1] == 1:
+                return 1
+
+        fixed_settings = np.array([[0, 7, 1], [5, 7, 2], [1, 0, 3], [5, 1, 1]])
+
+        spy_prod = mocker.spy(np, "prod")
+        spy_hstack = mocker.spy(np, "hstack")
+
+        postprocessed = qcut.qcut_processing_fn_mc(
+            fixed_samples, communication_graph, fixed_settings, shots, fn
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:2045]:
```suggestion
            cast_fixed_samples, communication_graph, fixed_settings, shots, fn
```

**Review by trbromley — COMMENTED** [review]
Thanks @anthayes92. I've left a final round of comments but should be good to merge once resolved.

**Comment by anthayes92** [comment]
Thanks for the suggestions @trbromley . Updates to the tests have been manually checked and added here: ceb13836b334dbb91fdb3a10f446c2aa2ba38635

File: pennylane/transforms/qcut.py [hunk 20]
```diff
@@ -755,45 +760,37 @@ def expand_fragment_tapes_mc(
...
+        ...     print(c2.draw())
+        ...     print("")
         ...
 
         config 0:
-        0: ──H──╭C──┤ Sample[Projector(M0)]
-        1: ─────╰X──┤ Sample[X]
-        M0 =
-        [1]
+        0: ──H─╭C─┤  Sample[|1⟩⟨1|]
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py:775]: kudos to the mystery contributor! :eyes:

**Review by trbromley — APPROVED** [review]
Thanks @anthayes92!

**Pull request merged**